### PR TITLE
I completed all task :

### PR DESCRIPTION
**Pull request by Thaithaothanhthu**
TASK 1:  user can't add a todo without a description.
TASK 2: user can mark a todo as completed.
Write a database migration script in resources/
TASK 3:  user can view a todo in a JSON format.
Ex: /todo/{id}/json => {id: 1, user_id: 1, description: "Lorem Ipsum"}
TASK 4: usercan see a confirmation message when I add/delete a todo.
TASK 5: user can see my list of todos paginated.
TASK 6: Implement own my ORM database access layer so we don’t have SQL in the controller code.

### CODE REVIEW

File: Entities/todo.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * @Entity @Table(name="products")
+ */
+class Todo
+{
+    /** @Id @Column(type="integer") @GeneratedValue */
+    protected $id;
+     /** @Id @Column(type="integer")*/
+    protected $user_id;
+	/** @Column(type="string") */
+	protected $description;
```
Comment (T0RAT0RA) [inline Entities/todo.php]: Indentation

File: Entities/user.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * @Entity @Table(name="products")
+ */
+class User
+{
+    /** @Id @Column(type="integer") @GeneratedValue */
+    protected $id;
+    	/** @Column(type="string") */
+    protected $username;
+	/** @Column(type="string") */
+	protected $password;
```
Comment (T0RAT0RA) [inline Entities/user.php]: Indentation

File: Entities/user.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * @Entity @Table(name="products")
```
Comment (T0RAT0RA) [inline Entities/user.php]: Wrong table name

File: src/controllers.php [hunk 4]
```diff
@@ -9,81 +10,127 @@
     return $twig;
 }));
 
+$before =  function (Request $request) use ($app) {
```
Comment (T0RAT0RA) [inline src/controllers.php]: I like that, but the variable could have a better name like `isAuthenticated`.

File: src/controllers.php [hunk 5]
```diff
@@ -9,81 +10,127 @@
...
-
+  
+    $user = $app['session']->get('user');
     if ($id){
-        $sql = "SELECT * FROM todos WHERE id = '$id'";
-        $todo = $app['db']->fetchAssoc($sql);
-
+      
+        $todo =Todo::getTodo($app,$id,$user['id']);
+        if($todo)
```
Comment (T0RAT0RA) [inline src/controllers.php]: `if` condition without curly bracket is generally a bad practice and can easilly lead to error.
I recommand to always surround your condition with `{ }`

File: src/controllers.php [hunk 6]
```diff
@@ -9,81 +10,127 @@
...
-        $sql = "SELECT * FROM todos WHERE id = '$id'";
-        $todo = $app['db']->fetchAssoc($sql);
-
+      
+        $todo =Todo::getTodo($app,$id,$user['id']);
+        if($todo)   
         return $app['twig']->render('todo.html', [
             'todo' => $todo,
         ]);
+        return "Can not access";
```
Comment (T0RAT0RA) [inline src/controllers.php]: Why not returning a 404 like the `'/todo/{id}/json'` endpoint?

File: templates/layout.html [hunk 7]
```diff
@@ -1,30 +1,41 @@
+<?php
+header("Expires: Tue, 01 Jan 2000 00:00:00 GMT");
```
Comment (T0RAT0RA) [inline templates/layout.html]: What is that??

File: templates/layout.html [hunk 8]
```diff
@@ -39,5 +50,15 @@
...
     <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
     <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js"></script>
+    <script type="text/javascript" src="{{ app.request.basepath }}/js/myscript.js"></script>
+    
+
+    <script type="text/javascript" charset="utf8" src="//cdn.datatables.net/1.10.13/js/jquery.dataTables.js"></script>
   </body>
+
+  <script type="text/javascript">
+
```
Comment (T0RAT0RA) [inline templates/layout.html]: Empty javascript?

File: web/.htaccess [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<IfModule mod_rewrite.c>
+    Options -MultiViews
```
Comment (T0RAT0RA) [inline web/.htaccess]: Why this change?

File: Entities/todo.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    {
+       $app['db']->insert('todos', array(
+        'user_id' => $user_id,
+        'description' => $description,
+            ));
+    }
+
+      public static function deleteTodo ($app,$id,$user_id)
+    {
+        $sql = "DELETE FROM todos WHERE id = '$id' AND user_id = '$user_id' ";
```
Comment (T0RAT0RA) [inline Entities/todo.php]: Why do you need the `user_id`?
You could access it through `$this->user_id`

File: Entities/todo.php (hunk 1 above)
Comment (Thaithaothanhthu) [inline Entities/todo.php]: you are right

File: Entities/todo.php (hunk 10 above)
Comment (Thaithaothanhthu) [inline Entities/todo.php]: because I created a static function, It can not access to the local properties of the class

File: Entities/user.php (hunk 3 above)
Comment (Thaithaothanhthu) [inline Entities/user.php]: you are right

File: Entities/user.php (hunk 2 above)
Comment (Thaithaothanhthu) [inline Entities/user.php]: you are right

File: src/controllers.php (hunk 4 above)
Comment (Thaithaothanhthu) [inline src/controllers.php]: you are right. thank you

File: src/controllers.php (hunk 5 above)
Comment (Thaithaothanhthu) [inline src/controllers.php]: you are right! Thank you

File: src/controllers.php (hunk 6 above)
Comment (Thaithaothanhthu) [inline src/controllers.php]: you are right! Thank you

File: templates/layout.html (hunk 7 above)
Comment (Thaithaothanhthu) [inline templates/layout.html]: this one I want to add some securities for disable caching in all browsers, because maybe the user close the browser suddenly without log out, and another person open the browser and he can go to that account without login. so i clear all the cache from browser.

File: templates/layout.html (hunk 8 above)
Comment (Thaithaothanhthu) [inline templates/layout.html]: you are right. I moved the code to another file (myscript.js) but i forgot to delete that code

File: web/.htaccess (hunk 9 above)
Comment (Thaithaothanhthu) [inline web/.htaccess]: because I used base_url in the controller so i uncommended:  RewriteBase /backend-php-test/web

**Comment by Thaithaothanhthu** [comment]
thank you for your time to review. and have a nice day
